### PR TITLE
keepassxc: Avoid absolute path to fix autostart

### PR DESCRIPTION
**Pull request by klemensn**
```
Tools > Settings > General > Basic Settings > Startup
[x] Automatically launch KeePassXC at system startup
```

If set, this will create ~/.config/autostart/org.keepassxc.KeePassXC.desktop
```
[Desktop Entry]
Name=KeePassXC
GenericName=Password Manager
Exec=/nix/store/jn5pxhskvy08rnq4yfbcrsmzjpiwj09h-keepassxc-2.7.1/bin/.keepassxc-wrapped
TryExec=/nix/store/jn5pxhskvy08rnq4yfbcrsmzjpiwj09h-keepassxc-2.7.1/bin/.keepassxc-wrapped
Icon=keepassxc
StartupWMClass=keepassxc
StartupNotify=true
Terminal=false
Type=Application
Version=1.0
Categories=Utility;Security;Qt;
MimeType=application/x-keepass2;
X-GNOME-Autostart-enabled=true
X-GNOME-Autostart-Delay=2
X-KDE-autostart-after=panel
X-LXQt-Need-Tray=true
```

I noticed this by seeing `..keepassxc-wrapped-wrapped` in my process
list long after having fixed double wrapping, so grepping the home
directory turned up this persisted state.

Apply my upstream PR to effectively replace the absoloute path with the
program name.

To update the config file, unset above option, press OK, renavigate to
it, set it and press OK again.

@jonafato @turion

### CODE REVIEW

**Comment by klemensn** [comment]
For reference: https://github.com/keepassxreboot/keepassxc/pull/8191

File: pkgs/applications/misc/keepassx/community.nix
Comment (turion) [inline pkgs/applications/misc/keepassx/community.nix:62]:
```suggestion
      # autostart: Linux: Exec= filename not absolute path
```

**Review by turion — APPROVED** [review]

File: pkgs/applications/misc/keepassx/community.nix
Comment (klemensn) [reply pkgs/applications/misc/keepassx/community.nix]: Fixed also in the commit message, thanks.

**Review by turion — CHANGES_REQUESTED** [review]
It seems this broke the integration test? Can you have a quick look into that? I assume it works fine on your machine?

**Comment by klemensn** [comment]
> It seems this broke the integration test? Can you have a quick look into that? I assume it works fine on your machine?

It most certainly works for me under KDE/Plasma, but I'll check out the tests now.

**Comment by klemensn** [comment]
> It seems this broke the integration test? Can you have a quick look into that? I assume it works fine on your machine?

Nope, `cd nixos/tests && nix-build ./keepassxc.nix` also fails without my change, so this is unrelated.

**Comment by klemensn** [comment]
Here are the integration test failures with 2b5b8b4, so whatever broke integration tests is in `master` already.
https://github.com/NixOS/nixpkgs/pull/178880/checks?check_run_id=7053450737
https://github.com/NixOS/nixpkgs/pull/178880/checks?check_run_id=7053444819

File: pkgs/applications/misc/keepassx/community.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/keepassx/community.nix:63]:
```suggestion
    # autostart: Linux: Exec= filename not absolute path
    (fetchpatch {
      url = "https://github.com/keepassxreboot/keepassxc/commit/4e0a2cd39ddbdb30b9308494381e6fb494ca8481.patch";
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/applications/misc/keepassx/community.nix
Comment (klemensn) [reply pkgs/applications/misc/keepassx/community.nix:63]: `pull/8191` makes it clear that this has not yet been committed.
`commit/4e0a2cd...` only works because GitHub makes commits available across repositories, even when my `4e0a2cd` has not yet been committed to `keepassxreboot/keepassxc` proper.

I've also put the comment inside the `fetchpatch` block because it
a) is immediately concerned about/attached to the following `url = ...` line (that's the title of the PR, hence using `pull/8191` and not a commit),
b) provides slightly better ergonomics for future hackers removing this patch on the next update (`da(` in vim conveniently zaps the entire block regardless of where the cursor is inside the parentheses).

If you strongly prefer a commit hash to a PR number, I'll wait for upstream to merge to be able to use their genuine commit hash.

Comment (SuperSandro2000) [reply pkgs/applications/misc/keepassx/community.nix:63]: > `pull/8191` makes it clear that this has not yet been committed.

and will break if you push to the branch or rebase it.



> I've also put the comment inside the `fetchpatch` block because it

it is convention that especial for CVE fixes the comment is put before the patch which might be vendored or fetched.



> If you strongly prefer a commit hash to a PR number, I'll wait for upstream to merge to be able to use their genuine commit hash.

It must be some hash. Using only the PR number is not allowed. If it is merged or not does not really matter. Even https://github.com/keepassxreboot/keepassxc/pull/8191/commits/4e0a2cd39ddbdb30b9308494381e6fb494ca8481.patch is allowed which redirects to what I suggested.

Comment (klemensn) [reply pkgs/applications/misc/keepassx/community.nix:63]: > > `pull/8191` makes it clear that this has not yet been committed.
> 
> and will break if you push to the branch or rebase it.

Only if the actual diff changes.
`fetchpatch` normalises before checksumming, so merely amending the commit message or rebasing where the changed file did not change will still result in the same output.

> it is convention that especial for CVE fixes the comment is put before the patch which might be vendored or fetched.
> 
> > If you strongly prefer a commit hash to a PR number, I'll wait for upstream to merge to be able to use their genuine commit hash.
> 
> It must be some hash. Using only the PR number is not allowed. If it is merged or not does not really matter. Even https://github.com/keepassxreboot/keepassxc/pull/8191/commits/4e0a2cd39ddbdb30b9308494381e6fb494ca8481.patch is allowed which redirects to what I suggested.

`git grep -E '/pull/[0-9]+\.(diff|patch)'` made the impression on me that this is in fact allowed in nixpkgs.

I like the `pull/.../commit/...` link. thanks.

Comment (SuperSandro2000) [reply pkgs/applications/misc/keepassx/community.nix]: > `git grep -E '/pull/[0-9]+\.(diff|patch)'` made the impression on me that this is in fact allowed in nixpkgs.

grepping through source code does not always reflect current policies as many old things are not updated within a reasonable timeframe.

**Comment by SuperSandro2000** [comment]
> > It seems this broke the integration test? Can you have a quick look into that? I assume it works fine on your machine?
> 
> Nope, `cd nixos/tests && nix-build ./keepassxc.nix` also fails without my change, so this is unrelated.

I am not sure if we want to merge this with broken tests. I would appreciate it if those where fixed, too.

**Comment by klemensn** [comment]
> > > It seems this broke the integration test? Can you have a quick look into that? I assume it works fine on your machine?
> > 
> > 
> > Nope, `cd nixos/tests && nix-build ./keepassxc.nix` also fails without my change, so this is unrelated.
> 
> I am not sure if we want to merge this with broken tests. I would appreciate it if those where fixed, too.

Sure, I'd appreciate that as well, but regardless of the fix I propose.

Tests are already broken without my change and blocking this PR seems counter-productive and unfair.
I would just stop contributing if I'd be required to fix someone else's regressions before fixing stuff I'm interested in.

File: pkgs/applications/misc/keepassx/community.nix
Comment (turion) [reply pkgs/applications/misc/keepassx/community.nix]: > I like the pull/.../commit/... link. thanks.

Yes, that's a good solution.

**Comment by turion** [comment]
Agree, it's really not your burden to fix a test you didn't break. Thanks for checking that you in fact didn't. I'll have a look into the integration test.

**Comment by SuperSandro2000** [comment]
I took the opportunity to do some research and reverted the last change to keepassxc (https://github.com/NixOS/nixpkgs/commit/1daf0125afd9c44723d16f6b3dad5fcbc3cde858). When reverting the linked commit the tests work. With the commit the tests are broken.

klemensn you are the author of that commit, so it is not someone else regression and I would appreciate it if you could take a look why the tests are broken and try to fix them.

**Comment by klemensn** [comment]
> I took the opportunity to do some research and reverted the last change to keepassxc (https://github.com/NixOS/nixpkgs/commit/1daf0125afd9c44723d16f6b3dad5fcbc3cde858). When reverting the linked commit the tests work. With the commit the tests are broken.

Now that's a surprise! Thanks.

> klemensn you are the author of that commit, so it is not someone else regression and I would appreciate it if you could take a look why the tests are broken and try to fix them.

I'll do that once I'm at a real computer again in a few days.

**Comment by klemensn** [comment]
> I'll do that once I'm at a real computer again in a few days.

I looked into the test failure but did not find out why it breaks.
I'll open a separate PR to revert #175104.

This PR has no effect on the nixpkg test, so I should be good to merge both before and after the revert.

**Comment by SuperSandro2000** [comment]
@ofborg eval

**Comment by turion** [comment]
> I'll open a separate PR to revert https://github.com/NixOS/nixpkgs/pull/175104.

@klemensn any progress on that?

**Comment by ilkecan** [comment]
I think this can be closed with #197411 as the referenced PR is accepted by upstream.

**Comment by bjornfor** [comment]
> I think this can be closed with https://github.com/NixOS/nixpkgs/pull/197411 as the referenced PR is accepted by upstream.

Closing.

**Pull request closed without merging**
